rest/handler.go

```diff
@@ -283,27 +283,51 @@ func (h *handler) logStatus(status int, message string) {
 	h.logDuration(false) // don't track actual time
 }
 
-func (h *handler) checkAuth(context *db.DatabaseContext) error {
-	h.user = nil
-	if context == nil {
-		return nil
+func (h *handler) statAuthSuccess() {
+	if h.db == nil || h.db.DatabaseContext == nil || h.db.DatabaseContext.DbStats == nil {
+		return
+	}
+	h.db.DatabaseContext.DbStats.StatsSecurity().Add(base.StatKeyAuthSuccessCount, 1)
+}
+
+func (h *handler) statAuthFailed() {
+	if h.db == nil || h.db.DatabaseContext == nil || h.db.DatabaseContext.DbStats == nil {
+		return
 	}
+	h.db.DatabaseContext.DbStats.StatsSecurity().Add(base.StatKeyAuthFailedCount, 1)
+}
+
+func (h *handler) checkAuth(context *db.DatabaseContext) (err error) {
+
+	// Record stats
+	defer func() {
+		if err != nil {
+			h.statAuthFailed()
+		} else {
+			h.statAuthSuccess()
+		}
+	}()
 
 	// Record TotalAuthTime stat
 	defer func(t time.Time) {
-		if context.DbStats == nil {
+		if context == nil || context.DbStats == nil {
```

Comment:
Similarly, move this below the context check below to avoid having to do the nil check again.   

In most usages we're assuming that DbStats isn't nil.  If that's potentially nil, we should refactor to avoid every caller having to do that check.  (probably by making context.DbStats a DatabaseStats instead of *DatabaseStats, and doing the nil check inside all the stat group accessors).

Reply:
Updated per above comment